New visualization search selection

maryia-lapata opened the pull request:
EUIfication of Wizard Step 2


Part of elastic/kibana#29608 


----


![feb-14-2019 14-21-45](https://user-images.githubusercontent.com/11524637/52783753-e8f86800-3063-11e9-85ad-6416a361ce3e.gif)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11268/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11308/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11517/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11564/)

cchaos commented:
Hey @maryia-lapata , great work so far, and thank you for converting this piece!

Based on your screenshots, the one problem I'm seeing is that not all the selections are `searches`. You're either choosing an index pattern **or** a saved search. So the wording should be more like:

<img src="https://d.pr/free/i/6FBbF9+" />

Also, can we get rid of the modal graphic (bottom right) for this modal and just keep it on the vis type selection modal?

maryia-lapata commented:
Hi @cchaos, thank you for your comment! I updated the PR and added a gif to the description. Any other suggestions are welcome.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11866/)

cchaos commented:
Looks great, ty. The only other thing I'd do is make the "Manage saved searches" button a hollow one by removing the `fill` prop because it's not the primary action. The primary action is selecting a saved search.

We may also want to consider the new [EuiListGroup](https://elastic.github.io/eui/#/display/list-group) instead of a table because it allows the whole row/item to be clicked instead of just the link and has better hover states. Tables should really only be used for real tabular data, when there are multiple columns not for selection.

timroes commented:
@cchaos I actually request a non hollow button. I would be fine with a white background button (do we have that?) But if you make it a hollow button the background image was visible through the button in the earlier screenshots which looked very weird.

With regards to the `EuiListGroup` that's an awesome new component I wasn't aware of. We'll make sure to use this, but I want to keep that separate from that PR. Actually we have one component (saved_object_finder) that is used for the dialog and also on the dashboard panel (since I saw you also commented on Joe's PR there), so we just need to exchange it in that component. But since that's a bit more work if there is no support for pagination in that component, we should keep that in a separate PR.

flash1293 commented:
@cchaos In this issue https://github.com/elastic/kibana/issues/27985 we are planning to get rid of the tabs in favor of a unified list (not decided yet). Should we move this requirement over to this issue since we plan to modify this component anyway?

timroes commented:
Created #30532 for the component replace task.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12205/)

maryia-lapata commented:
@cchaos , @timroes I removed the `fill` prop from the "Manage saved searches" button.
@timroes  I think it looks better then at the first time, I updated the description, you can take a look

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12243/)

cchaos commented on src/legacy/core_plugins/kibana/public/visualize/wizard/_dialog.scss:
```diff
@@ -5,6 +5,11 @@
 
+.visNewVisSearchDialog {
+  min-height: 480px;
+  max-width: 810px;
```
The width of this one doesn't need to match the chart type selection and since modal's already have a max-width, this can be removed.

cchaos commented on src/legacy/core_plugins/kibana/public/visualize/wizard/_dialog.scss:
```diff
@@ -5,6 +5,11 @@
 }
 
+.visNewVisSearchDialog {
+  min-height: 480px;
```
Can you use SASS variables instead of `px` values? For instance this should be `$euiSizeL * 20`

timroes commented:
>The search bar for the index patterns doesn't seem flexible enough to search any part of the index pattern name. For instance I tried typing "da" to match "kibana_sample_**da**ta_ecommerce", but came up with no results:

That's currently implemented in the API of saved objects like that and happens because the API currently only searches for word beginnings (since it's way cheaper in ES), and the underscore is a wordchar, meaning that index patterns are a single word, and you cannot search in the middle for it. I don't think anything we need to address right now, and we really would need to think carefully about allowing prefix wildcards here ,since it's way more expensive then suffix wildcards.

> Is there any way to tap into the Saved objects page to pre-fill or auto-select the "search" object type so that clicking the "Manage saved searches" doesn't just go to a list of all saved objects?

Unfortunately currently there isn't as far as I remember (cc @jen-huang I think I asked you the same at some point in time?) we could create a separate issue that allows that, but I would rather like not having it all solved in this PR, for the sake of reviewable PRs :)

timroes commented:
Also we just discussed offline, that it might make sense to add the "Manage Index Pattern" button too, since we now have enough space for it, and I don't see a good reason why we should have only one manage button there for searches but not for index patterns. So I rather would like both to have that button or none.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12574/)

cchaos commented:
Haha, @timroes I was actually going to suggest the opposite (remove the "Manage" buttons).

My reasoning is that since the "Manage Saved Searches" only brings you to the full saved objects list and really only allows you to delete it, it doesn't seem super helpful at the point when users are trying to create a visualization. I would think that they'd probably want to "Create" a new saved search if they don't find what they're looking for in the list. Though that would require just sending them back to discover, and if they've never created a saved search before, they'll have no clue what to do next (unless we can tap into some routing mechanism that allows a modal to pop up when they land on discover explaining to them what to do next).

For index patterns, and maybe @AlonaNadler could help based on customers, but I don't see any benefit to passing them to create a new index pattern during the visualization build process.

AlonaNadler commented:
@cchaos looking at the first gif for index pattern in Visualize, how many index pattern will show up by default? also why in @maryia-lapata gif there is scrolling in the chart type screen? 
Regarding the `manage saved objects` or `manage index pattern` I honestly not sure how helpful it is when you are in the process of creating a new visualization. The common confusion is which index pattern/saved search to choose,  if that is the question, users will end up in discover looking at what's in each index pattern to figure which one they need.

Crazybus commented:
retest

maryia-lapata commented:
@cchaos the "Manage" buttons were removed.

@AlonaNadler regarding to scrolling, since `SavedObjectFinder`, which is used on the modal, will be updated to use `EuiListGroup` (#30532), I'll get back to this matter later.
For now we agreed with @timroes to show 8 items per page (5 is too little and 10 is too many). 

---
Maybe it would be helpful for the user to have `back` button to change a vis type. So I added pencil icon near the header.
![image](https://user-images.githubusercontent.com/11524637/52723393-7bdec700-2fbe-11e9-8496-53692588ef1b.png)

 As an option it can be text `change`  instead of icon

![changbtn](https://user-images.githubusercontent.com/11524637/52723165-1094f500-2fbe-11e9-898c-5f859696a745.png)

What do you think about `back` button?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/266/)

cchaos commented:
I'm a bit on the fence about a back button because if they decide they want to start over they can just click the (x) and then click "New" again. However, if we think a back button is necessary, I would actually Make the modal header more like breadcrumbs and make the whole arrow and chart type part a link back to the first modal like so:

![source](https://user-images.githubusercontent.com/549577/52724001-8e381100-2f7c-11e9-84ab-005faa5c6f35.png)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/276/)

timroes commented:
@cchaos and I talked offline, and we think we should leave the back button for now. It should not provide too much benefit, since going back at that point might not be very likely, so we'll go for the first version with it, until we're figuring out that there is a need for it.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/498/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/532/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/546/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/824/)

flash1293 commented on src/legacy/ui/public/saved_objects/components/saved_object_finder.tsx:
```diff
@@ -52,31 +51,38 @@ interface SavedObjectFinderUIState {
+
+interface SavedObjectFinderFixedPage extends BaseSavedObjectFinder {
+  initialPageSize?: undefined;
+  fixedPageSize: number;
```
This seems more complicated than necessary - why not adding another prop `hidePerPageOptions` which defaults to `false`?

timroes replied:
We had that solution beforehand. It actually doesn't ease the code here at all, since we still must make sure, once that flag is set, that `initialPageSize` can be any value, but only can be `5 | 10 | 15` in case that flag is not set. So it wouldn't really change a thing here, and I found the wording that way a bit more explicit, that you know you're specifying the actual page size and not just an initial one.

flash1293 replied:
Makes sense, thanks. TBH I think we should just get rid of the page size dropdown for this component altogether.

timroes replied:
I think once we're switching over to using `EuiListGroup` we can anyway reevaluate what we need in it :)

flash1293 commented on src/legacy/core_plugins/kibana/public/visualize/wizard/new_vis_modal.tsx:
```diff
@@ -45,33 +53,65 @@ class NewVisModal extends React.Component<TypeSelectionProps> {
+  private onSearchSelected = (searchId: string, searchType: string) => {
     this.props.onClose();
+
+    if (!this.state.visType) {
```
This check shouldn't be necessary - line 69 only renders the SearchSelection if `this.state.visType` is set. Am I missing something?

maryia-lapata replied:
Since `visType` is optional (it's defined after vis type selection by the user), `tslint` warns at line [107](https://github.com/elastic/kibana/pull/30196/files#diff-a52663fb9e8ccb0ee2abd6f8c8cded68R107). 
Since `onSearchSelected` will be invoked when `visType` is already defined, I removed this check and used a non-null assertion at 107 line

flash1293 commented on src/legacy/core_plugins/kibana/public/visualize/wizard/search_selection/search_selection.tsx:
```diff
@@ -0,0 +1,148 @@
+  constructor(props: SearchSelectionProps) {
+    super(props);
+
+    this.tabs = [
```
I know the add panel does this as well, but AFAIK and IMHO the SavedObject finder should always be initialized as part of the render function - if we are doing it in the constructor, the definitions can get stale (e.g. when a prop changes, it won't re-render correctly).

https://stackoverflow.com/questions/47875097/add-element-to-a-state-react

maryia-lapata replied:
Thank you for the comment. I updated `SearchSelection`.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1183/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1257/)

timroes commented on src/legacy/ui/public/saved_objects/components/saved_object_finder.tsx:
```diff
@@ -52,31 +51,38 @@ interface SavedObjectFinderUIState {
+    savedObjectType: PropTypes.oneOf(['visualization', 'search', 'index-pattern']).isRequired,
     visTypes: PropTypes.object,
-    initialPageSize: PropTypes.number,
+    pagination: PropTypes.object,
```
I think this `pagination` prop type is now wrong. We could either simply use:

```ts
initialPageSize: PropTypes.oneOf([5, 10, 15]),
fixedPageSize: PropTypes.number,
```

Alternatively there is a more complex way of writing a custom prop validator that really checks if only one of it is set, but since we're switching over to TS in most places I don't think that's required for now.

elasticmachine commented:
Pinging @elastic/kibana-app

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1267/)
